R/check_exercise.R

```diff
@@ -0,0 +1,407 @@
+
+  }
+
+  # Setting something to return that will be turned to FALSE
+  # if any check fail
+  all_passed <- TRUE
+
+  # Now that we have manipulated the chunks, we can build the
+  # exercise objects
+
+  for (chunk_ in chunks_){
+    # Restoring  knitr code stock
+    knitr::knit_code$restore()
+    # Grep all the related chunks (i.e setup, checker, etc)
+    all_related <- grep(chunk_, names(usefull_chunks), value = TRUE)
+
+    # If there is a setup chunk, we make sure it's the first of the list
+    if (any(grepl("setup", all_related))){
+      setup_ <- grepl("setup", all_related)
+      all_related <- c(
+        all_related[setup_],
+        all_related[!setup_]
+      )
+    }
+
+    # This function allow to grab a chunk based on
+    # its pattern
+    grab_chunk <- function(
+      pattern
+    ){
+      res <- usefull_chunks[
+        all_related[
+          grepl(
+            sprintf("%s-%s", chunk_, pattern),
+            all_related
+          )
+        ]
+      ]
+      if (length(res)){
+        return(res)
+      }
+      list(
+        list()
+      )
+    }
+
+    # Now we are building the exercise object, and it
+    # will be sent to evaluate_code()
+    #
+    # An exercise object needs the following elements:
+    # - exercise$label
+    # - exercise$code, which will be the solution code here
+    # - exercise$restore
+    # - exercise$timestamp
+    # - exercise$global_setup if any
+    # - exercise$setup if any
+    # - exercise$code_check if any
+    # - exercise$chunks that contains the setup chunks
+    # - exercise$check if any
+    # - exercise$engine
+    #
+
+    # We keep the  chunk we need: setup, and solution
+    chunks_needed <- list()
+
+    if (length(grab_chunk("setup")[[1]])){
+      chunks_needed[[1]] <- grab_chunk("setup")[[1]]
+    }
+    if (length(grab_chunk("solution")[[1]])){
+      chunks_needed[[
+        length(chunks_needed) + 1
+      ]] <- grab_chunk("solution")[[1]]
+    }
+    exercise_blt <- list(
+      # Getting the label
+      label = {
+        chunk_
+      },
+      code = {
+        paste0(grab_chunk("solution")[[1]]$code, "\n\n", collapse = "\n")
+      },
+      restore = {
+        FALSE
+      },
+      timestamp = {
+        as.numeric(Sys.time())
+      },
+      global_setup = {
+        paste0(setup_chunk$code, collapse = "\n")
+      },
+      setup = {
+        paste0(grab_chunk("setup")[[1]]$code, collapse = "\n")
+      },
+      chunks = {
+        chunks_needed
+      },
+      solution  = {
+        grab_chunk("solution")[[1]]
+      },
+      code_check  = {
+        # TODO
+        #grab_chunk("code-check")[[1]]
+      },
+      options = {
+        usefull_chunks[chunk_][[1]]$options
+      },
+      engine = {
+        usefull_chunks[chunk_][[1]]$enginee %||% "r"
+      },
+      version = "1"
+    )
+
+    res <- evaluate_exercise(
+      exercise_blt,
+      envir = new.env()
```

Comment:
Use the previously source `'setup'` environment

```suggestion
      envir = setup_chunk_envir
```